crates/sui-data-ingestion/src/worker_pool.rs

```diff
@@ -0,0 +1,108 @@
+        Self {
+            task_name: worker.name().to_string(),
+            concurrency,
+            worker: Arc::new(worker),
+        }
+    }
+    pub async fn run(
+        self,
+        mut current_checkpoint_number: CheckpointSequenceNumber,
+        mut checkpoint_receiver: broadcast::Receiver<CheckpointData>,
+        executor_progress_sender: mpsc::Sender<(String, CheckpointSequenceNumber)>,
+    ) {
+        info!(
+            "Starting indexing pipeline {} with concurrency {}",
+            self.task_name, self.concurrency
+        );
+        let mut updates: HashSet<u64> = HashSet::new();
+
+        let (progress_sender, mut progress_receiver) = mpsc::channel(MAX_CHECKPOINTS_IN_PROGRESS);
+        let mut workers = vec![];
+
+        // spawn child workers
+        for _ in 0..self.concurrency {
+            let (worker_sender, mut worker_recv) =
+                mpsc::channel::<CheckpointData>(MAX_CHECKPOINTS_IN_PROGRESS);
+            let (term_sender, mut term_receiver) = oneshot::channel::<()>();
+            let cloned_progress_sender = progress_sender.clone();
+            workers.push((worker_sender, term_sender));
+
+            let worker = self.worker.clone();
+            spawn_monitored_task!(async move {
+                loop {
+                    tokio::select! {
+                        _ = &mut term_receiver => break,
+                        Some(checkpoint) = worker_recv.recv() => {
+                            let backoff = backoff::ExponentialBackoff::default();
+                            backoff::future::retry(backoff, || async {
+                                worker
+                                    .clone()
+                                    .process_checkpoint(checkpoint.clone())
+                                    .await
+                                    .map_err(backoff::Error::transient)
+                            })
+                            .await
+                            .expect("checkpoint processing failed for checkpoint");
+                            cloned_progress_sender
+                                .send(checkpoint.checkpoint_summary.sequence_number)
+                                .await
+                                .expect("failed to update progress");
+                        }
+                    }
+                }
+            });
+        }
+        // main worker pool loop
+        loop {
+            tokio::select! {
+                checkpoint_result = checkpoint_receiver.recv() => {
+                    match checkpoint_result {
+                        Ok(checkpoint) => {
+                            let sequence_number = checkpoint.checkpoint_summary.sequence_number;
+                            if sequence_number < current_checkpoint_number {
+                                continue;
+                            }
+                            let worker_id = (sequence_number % self.concurrency as u64) as usize;
```

Comment:
each workflow will have a separate worker pool, making them somewhat independent (though they share the same machine resources). However, each workflow has a different set of workers for processing, and each reports a separate watermark to the progress store. The local reader is somewhat impacted by the slowest workflow, as it can only prune local checkpoint file once every task is done with it.
The concurrency variable represents the number of workers for a given workflow. We can discuss the assignment mechanism more. Currently, worker_i processes checkpoints where checkpoint_number % num_of_workers_in_pool == i. This way, the slowest worker doesn't prevent its neighbors from processing more checkpoints and making a progress. However, the watermark is only updated once there's a continuous sequence from a previous checkpoint